File: tyhp-lang/phpstorm/src/unitTest/kotlin/com/tyhp/lang/settings/SettingsCoreTest.kt
package com.tyhp.lang.settings

import kotlin.test.Test
import kotlin.test.assertEquals
import kotlin.test.assertFalse
import kotlin.test.assertTrue

class SettingsCoreTest {
    @Test
    fun `empty and whitespace path values are unset`() {
        assertTrue(isPathUnset(null))
        assertTrue(isPathUnset(""))
        assertTrue(isPathUnset("   "))
        assertFalse(isPathUnset("/usr/local/bin/tyhp"))
    }

    @Test
    fun `path write target is application unless a project override exists`() {
        assertEquals(PathWriteTarget.Application, pathWriteTarget(null))
        assertEquals(PathWriteTarget.Application, pathWriteTarget(InspectedPath()))
        assertEquals(
            PathWriteTarget.Application,
            pathWriteTarget(InspectedPath(applicationValue = "/user/tyhp")),
        )
        assertEquals(
            PathWriteTarget.Project,
            pathWriteTarget(InspectedPath(projectValue = "")),
        )
        assertEquals(
            PathWriteTarget.Project,
            pathWriteTarget(InspectedPath(projectValue = "/ws/tyhp")),
        )
    }

    @Test
    fun `effective path prefers project override including empty string`() {
        assertEquals(
            "/user/tyhp",
            effectiveTyhpPath(InspectedPath(applicationValue = "/user/tyhp")),
        )
        assertEquals(
            "/ws/tyhp",
            effectiveTyhpPath(InspectedPath(applicationValue = "/user/tyhp", projectValue = "/ws/tyhp")),
        )
        assertEquals(
            "",
            effectiveTyhpPath(InspectedPath(applicationValue = "/user/tyhp", projectValue = "  ")),
        )
        assertTrue(isPathUnset(effectiveTyhpPath(InspectedPath(projectValue = ""))))
    }

    @Test
    fun `install mode parsing`() {
        assertEquals(InstallMode.PATH, parseInstallMode("path"))
        assertEquals(InstallMode.GLOBAL, parseInstallMode("global"))
        assertEquals(InstallMode.EXTENSION, parseInstallMode("extension"))
        assertEquals(InstallMode.PATH, parseInstallMode("nope"))
        assertEquals(InstallMode.PATH, parseInstallMode(null))
    }

    @Test
    fun `release tags normalize and compare with optional v prefix`() {
        assertEquals("v805.0.0-alpha.1", normalizeReleaseTag("805.0.0-alpha.1"))
        assertEquals("v805.0.0-alpha.1", normalizeReleaseTag("v805.0.0-alpha.1"))
        assertEquals("", normalizeReleaseTag(""))
        assertTrue(tagsMatch("805.0.0-alpha.1", "v805.0.0-alpha.1"))
        assertFalse(tagsMatch("v805.0.0-alpha.2", "v805.0.0-alpha.1"))
    }
}

File: tyhp-lang/vscode/src/config/settingsCore.test.ts
import assert from "node:assert/strict";
import { test } from "node:test";
import {
    SettingsScope,
    explicitInspectedValue,
    isPathUnset,
    normalizeReleaseTag,
    parseInstallMode,
    pathWriteTarget,
    tagsMatch,
} from "./settingsCore";

test("empty and whitespace path values are unset", () => {
    assert.equal(isPathUnset(undefined), true);
    assert.equal(isPathUnset(null), true);
    assert.equal(isPathUnset(""), true);
    assert.equal(isPathUnset("   "), true);
    assert.equal(isPathUnset("/usr/local/bin/tyhp"), false);
});

test("path write target is User/Global unless a workspace override exists", () => {
    assert.equal(pathWriteTarget(undefined), SettingsScope.Global);
    assert.equal(pathWriteTarget({}), SettingsScope.Global);
    assert.equal(pathWriteTarget({ globalValue: "/user/tyhp" }), SettingsScope.Global);
    assert.equal(pathWriteTarget({ workspaceValue: "" }), SettingsScope.Workspace);
    assert.equal(pathWriteTarget({ workspaceValue: "/ws/tyhp" }), SettingsScope.Workspace);
    assert.equal(
        pathWriteTarget({ workspaceFolderValue: "/folder/tyhp", workspaceValue: "/ws/tyhp" }),
        SettingsScope.WorkspaceFolder
    );
});

test("install mode parsing", () => {
    assert.equal(parseInstallMode("path"), "path");
    assert.equal(parseInstallMode("global"), "global");
    assert.equal(parseInstallMode("extension"), "extension");
    assert.equal(parseInstallMode("nope"), "path");
    assert.equal(parseInstallMode(undefined), "path");
});

test("release tags normalize and compare with optional v prefix", () => {
    assert.equal(normalizeReleaseTag("805.0.0-alpha.1"), "v805.0.0-alpha.1");
    assert.equal(normalizeReleaseTag("v805.0.0-alpha.1"), "v805.0.0-alpha.1");
    assert.equal(normalizeReleaseTag(""), "");
    assert.equal(tagsMatch("805.0.0-alpha.1", "v805.0.0-alpha.1"), true);
    assert.equal(tagsMatch("v805.0.0-alpha.2", "v805.0.0-alpha.1"), false);
});

test("explicit inspected value prefers folder then workspace then user", () => {
    assert.equal(explicitInspectedValue(undefined), undefined);
    assert.equal(explicitInspectedValue({}), undefined);
    assert.equal(explicitInspectedValue({}), undefined);
    assert.equal(explicitInspectedValue({ globalValue: 9005 }), 9005);
    assert.equal(explicitInspectedValue({ globalValue: 9005, workspaceValue: 9010 }), 9010);
    assert.equal(
        explicitInspectedValue({
            globalValue: 9005,
            workspaceValue: 9010,
            workspaceFolderValue: 9111,
        }),
        9111
    );
});
